### PR TITLE
Translation improvements

### PR DESCRIPTION
**Pull request by zml2008**
One more step on top of @lucko's work, to restrict conflicting registrations.

### CODE REVIEW

File: api/src/main/java/net/kyori/adventure/translation/TranslationSource.java [hunk 1]
```diff
@@ -33,8 +35,17 @@
...
 public interface TranslationSource {
+  /**
+   * A key identifying this translation source.
+   *
+   * <p>Intended to be used for display to users.</p>
+   *
+   * @return an identifier for this translation source
+   * @since 4.0.0
+   */
+  Key name();
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/TranslationSource.java:47]:
```suggestion
  @NonNull Key name();
```

File: api/src/main/java/net/kyori/adventure/translation/TranslationSource.java [hunk 2]
```diff
@@ -44,4 +55,12 @@
...
    */
   @Nullable MessageFormat translate(final @NonNull String key, final @NonNull Locale locale);
+
+  /**
+   * Get the keys this source can translate.
+   *
+   * @return an unmodifiable collection containing this source's keys.
+   * @since 4.0.0
+   */
+  Collection<String> keys();
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/TranslationSource.java:65]:
```suggestion
  @NonNull Collection<String> keys();
```

File: api/src/main/java/net/kyori/adventure/translation/TranslationRegistryImpl.java [hunk 3]
```diff
@@ -52,19 +75,56 @@ public void unregister(final @NonNull String key) {
     this.translations.remove(key);
   }
 
+  @Override
+  public Key name() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/TranslationRegistryImpl.java:79]:
```suggestion
  public @NonNull Key name() {
```

File: api/src/main/java/net/kyori/adventure/translation/TranslationRegistryImpl.java [hunk 4]
```diff
@@ -52,19 +75,56 @@ public void unregister(final @NonNull String key) {
...
+
   @Override
   public @Nullable MessageFormat translate(final @NonNull String key, final @NonNull Locale locale) {
     final Translation translation = this.translations.get(key);
     if(translation == null) return null;
     return translation.translate(locale);
   }
 
+  @Override
+  public Collection<String> keys() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/TranslationRegistryImpl.java:91]:
```suggestion
  public @NonNull Collection<String> keys() {
```

File: api/src/main/java/net/kyori/adventure/translation/GlobalTranslationSourceImpl.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  @Override
+  public void unregister(final @NonNull TranslationSource source) {
+    requireNonNull(source, "source");
+    this.keys.values().remove(source);
+  }
+
+  @Override
+  public Key name() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/GlobalTranslationSourceImpl.java:71]:
```suggestion
  public @NonNull Key name() {
```

File: api/src/main/java/net/kyori/adventure/translation/GlobalTranslationSourceImpl.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    requireNonNull(key, "key");
+    requireNonNull(locale, "locale");
+    final TranslationSource source = this.keys.get(key);
+    if(source == null) return null;
+
+    return source.translate(key, locale);
+  }
+
+  @Override
+  public Collection<String> keys() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/GlobalTranslationSourceImpl.java:86]:
```suggestion
  public @NonNull Collection<String> keys() {
```

File: api/src/main/java/net/kyori/adventure/translation/GlobalTranslationSourceImpl.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  @Override
+  public void register(final @NonNull TranslationSource source) {
+    requireNonNull(source, "source");
+    if(source == this) {
+      throw new IllegalArgumentException("cannot register GlobalTranslationSource to itself");
+    }
+    for(final String key : source.keys()) {
+      final TranslationSource existing = this.keys.putIfAbsent(key, source);
+      if(existing != null && !existing.equals(source)) {
+        throw new IllegalArgumentException("Translation key '" + key + "' was found in both the " + source.name() + " and " + existing.name() + " translation sources!");
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/translation/GlobalTranslationSourceImpl.java:59]: Perhaps something like ``Cannot register translation key {key} from {source} as it has already been registered from {existing]`?

**Review by kashike — CHANGES_REQUESTED** [review]

**Review by lucko — APPROVED** [review]

**Review by kashike — APPROVED** [review]

**Pull request merged**